llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp

```diff
@@ -1120,6 +1122,129 @@ bool VPlanTransforms::handleMaxMinNumReductions(VPlan &Plan) {
+      Plan.getConstantInt(APInt::getMaxValue(Ty->getIntegerBitWidth()));
+  assert(FindIVPhiR->getVFScaleFactor() == 1 &&
+         "FindIV reduction must not be scaled");
+  ReductionStyle Style = RdxUnordered{1};
+  auto *FirstIdxPhiR = new VPReductionPHIRecipe(
+      dyn_cast_or_null<PHINode>(FindIVPhiR->getUnderlyingValue()),
+      RecurKind::UMin, *MaxInt, *FindIVPhiR->getBackedgeValue(), Style,
+      FindIVPhiR->hasUsesOutsideReductionChain());
+  FirstIdxPhiR->insertBefore(FindIVPhiR);
+
+  VPInstruction *FindLastIVResult =
+      findUserOf<VPInstruction::ComputeFindIVResult>(FindIVPhiR);
+  MinMaxResult->moveBefore(*FindLastIVResult->getParent(),
+                           FindLastIVResult->getIterator());
+
+  // The reduction using MinMaxPhiR needs adjusting to compute the correct
+  // result:
+  //  1. Find the first canonical indices corresponding to partial min/max
+  //     values, using loop reductions.
+  //  2. Find which of the partial min/max values are equal to the overall
+  //      min/max value.
+  //  3. Select among the canonical indices those corresponding to the overall
+  //     min/max value.
+  //  4. Find the first canonical index of overall min/max and scale it back to
+  //     the original IV using VPDerivedIVRecipe.
+  //  5. If the overall min/max is equal to the start value, the condition in
+  //  the
+  //     loop was always false, due to being strict; return the start value in
+  //     that case.
+  //
+  // The original reductions need adjusting:
+  // For example, this transforms
+  // vp<%min.result> = compute-reduction-result ir<%min.val>, ir<%min.val.next>
+  // vp<%find.iv.result> = compute-find-iv-result ir<%min.idx>, ir<0>,
+  //                                              ir<Sentinel>,
+  //                                              vp<%min.idx.next>
+  //
+  // into:
+  //  vp<%min.result> = compute-reduction-result ir<%min.val>, ir<%min.val.next>
+  //  vp<%final.min.cmp> = icmp eq ir<%min.val.next>, vp<%min.result>
+  //  vp<%final.min.idx> = select vp<%final.min.cmp>, ir<%min.idx.next>,
+  //  ir<MaxUInt> vp<%13> = compute-reduction-result ir<%min.idx>,
+  //  vp<%final.min.idx> vp<%scaled.result.iv> = DERIVED-IV ir<20> + vp<%13> *
+  //                                                        ir<1>
+  //  vp<%always.false> = icmp eq vp<%min.result>, ir<%original.min.start>
+  //  vp<%final.result> = select vp<%always.false>, vp<%scaled.result.iv>,
+  //                             ir<%original.start>
+
+  VPBuilder Builder(FindLastIVResult);
+  VPValue *MinMaxExiting = MinMaxResult->getOperand(1);
+  auto *FinalMinMaxCmp =
+      Builder.createICmp(CmpInst::ICMP_EQ, MinMaxExiting, MinMaxResult);
+  VPValue *LastIVExiting = FindLastIVResult->getOperand(3);
+  auto *FinalIVSelect =
+      Builder.createSelect(FinalMinMaxCmp, LastIVExiting, MaxInt);
+  VPSingleDefRecipe *FinalResult = Builder.createNaryOp(
+      VPInstruction::ComputeReductionResult, {FirstIdxPhiR, FinalIVSelect}, {},
```

Comment:
The original FindIV reduction phi `FindIVPhiR` is replaced by the UMin reduction phi `FindIdxPhiR`, in order to feed the simpler sentinel-less `ComputeReductionResult` here, than use the overly complex `ComputeFindIVResult`?
Could a `ComputeReductionResult` be formed here w/o a reduction phi feeding it with Kind/isInLoop/isOrdered (as its first operand), say by providing them directly via flags (at-least for non-fp reductions free of FMF)?

Reply:
Yep, I'll look into that separately, thanks. Not sure if flags will work as-is currently due to FMFs

Reply:
Them 9 FMFs are already combined with the 32 predicates of CmpInst::Predicate to form FCmpFlagsTy, so could follow by also combining FMFs with the <32 elements of RecurKind (needing only the 11 FP recurrence kinds w/o the trinary FAddMul which is for inloop only).